Book style tweak: add more suggestions in "CSS ≡"

poire-z opened the pull request:
As promised in https://github.com/koreader/koreader/pull/11353#issuecomment-1890929091.

Lots of text, may be I went too far... Dunno if we should bother translators with these bits of CSS tutorial...

Anyway, can be handy when needed, when you don't remember the syntax or a name, or just for exploring and getting ideas and experimenting.

I wasn't really inspired when organizing them into submenus... Thought about having them by topic (ie. Footnotes, DocFragments, Making pages), but most of these are not one-topic-centered, and can be used in varied contexts.

![image](https://github.com/koreader/koreader/assets/24273478/2c6d026d-daf4-4973-9c94-ecae1e944a06)

![image](https://github.com/koreader/koreader/assets/24273478/1f388106-beb7-4b11-8a01-ea5be6706f90)

![image](https://github.com/koreader/koreader/assets/24273478/9a4441d8-df63-455b-86d3-8139202ca74b)

![image](https://github.com/koreader/koreader/assets/24273478/d974b15a-52e2-4ed6-b22d-f4c8a9219111)

![image](https://github.com/koreader/koreader/assets/24273478/4ce25cf5-18fa-440b-9df8-36de218871a6)
(This `noteref-ignore` is not yet available - it just happened this weekend I needed that; so I could write a bit about this fresh use-case. May be I'll complete the descriptions of others when I'll meet again some classic use-cases.)

![image](https://github.com/koreader/koreader/assets/24273478/5d12a678-5ae9-44d1-a1dd-ffe18f92cd57)
(These ones are me planning for my old days, cf. https://github.com/koreader/crengine/pull/345#issuecomment-1673394857 :))

Pinging @jonnyl2 @Commodore64user as you recently had to play with Book style tweaks: would you feel these and my prose helpful? Also, feel free to review my wording, grammar, English, and inconsistencies (hard to stay consistent with so many little bits of mixed text and code).

@offset-torque : I don't think these needs to be duplicated into the User Guide (there may be better CSS introductions, and I wouldn't be able to write any - these bits expect people to already be familiar with "selectors", "properties", "element"), but you could mention this exists, somewhere.
Or maybe the 2 longer bits of text in the screenshots above (the use-case for Book style tweak, and the tip about View HTML) could end up somewhere.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/11533)
<!-- Reviewable:end -->

Frenzie commented on frontend/apps/reader/modules/readerstyletweak.lua:
```diff
@@ -783,13 +784,125 @@ local BOOK_TWEAK_INPUT_HINT = T([[
-    { "text-indent: 1.2em !important;", _("1.2em is our default text indentation.")},
+    { _("Long-press for info ⓘ"), _([[
+This popup menu provides a quick CSS cheat sheet, showing common selector syntax and classic CSS properties.
+It also shows some KOReader specific, non-standard, CSS features, that can be useful with e-books.
```
More a matter of taste, but I think it reads more fluid without the commas.
```suggestion
It also shows some KOReader-specific non-standard CSS features, that can be useful with e-books.
```

Frenzie commented:
It does seem a bit excessive to explain all of the basic CSS selectors, though I suppose not sending people to MDN does have the advantage of them not getting ideas like `|`. (That is, I assume the namespace selector isn't supported because you didn't include it; I didn't actually test it or look in the code.)

offset-torque commented:
**(Replying to my mention)**
I agree. These are pretty advanced things and duplicating in the guide won't provide extra benefits. For the guide I will mention this in the spirit of: **"We have a toolbox here if you want to fix things yourself"**.

poire-z commented:
> not getting ideas like `|`. (That is, I assume the namespace selector isn't supported because you didn't include it; I didn't actually test it or look in the code.)

(Although crengine XML parse remember the namespace of element and attributes, I think we just skip the namespace when parsing CSS and matching element and attributes. Which is mostly just fine and easier to not bother about them.)

jonnyl2 commented:
>Pinging @jonnyl2 (...) as you recently had to play with Book style tweaks: would you feel these and my prose helpful?

Yes, definitely helpful for inspiration to experiment and as shortcuts! I don't have anything to add or correct in regards to the help text.

Commodore64user commented:
I was going to make some comments here, but oh well, I will get back to you soon.

Commodore64user commented:
THIS IS CONSTRUCTIVE CRITICISM, please do not take this personal.

My first thought is: who is this useful for, the person who already knows CSS? no, he already knows what to do. The person who does not know CSS? perhaps, the main problem I see is: maybe, in trying to be too helpful, you are risking making this way too complicated for someone who does not know what any of that means. If it is only for when you might not remember exactly what you need, then the explanations are mostly unnecessary. 

Don't get me wrong, I think this is great, but maybe you should stop and consider, who is your target audience? then adjust the product accordingly. Because it seems to me, you are currently trying to cover all the positions with just one player.

I have also added a slight rewording for the second picture:

> This menu provides a non-exhaustive CSS syntax and properties list. It also shows some KOReader-specific, non-standard CSS features that can be used with your e-books. 
Most of these bits are already used by our categorised 'Style tweaks' (found in the top menu). Long-press on any style-tweak option to see its code and its expected results. Should these not be enough to achieve your desired look, you may need to adjust them slightly: tap once on the CSS code-box to copy the code to the clipboard, paste it here and edit it. 
Don't forget that you can long-press on any item to get more information on what it does and how it affect the look of your books.

are private properties the koreader-specific ones? if so, `In-house CSS properties` might be better

poire-z commented:
(No personalism taken :))

Well, the target audience is... myself. And people who knows CSS. There are some bits of CSS that are rarely used (ie `~` sibling selector, many `[attr??name]` attributes) and that I/we may have seen but not always remember 1) they exist and 2) the correct syntax. The main aim is to solve that.
Then, I could just wrote only a few words, ie. `~: any previous sibling`, but then, it would be helpful to those that once met and use them.
By writting a bit more, I still solve my use case, and _maybe_ it gives a bit more context to the people who are not confortable yet, and that may try it (so, enough info for them to guess how to try it), or go look on the web for more.
So, in the end, I need a bit of consistency in the simple and sufficient explanations for each of them.

Your wording is nice (I would remove the _Don't forget that you can_ because, even if long-press is common for that, it feels like a new information in that context :)). Fine with replacing that intro text with your wordings.

> are private properties the koreader-specific ones? if so, `In-house CSS properties` might be better

I think it's quite the usage to name any `-xyz-something` (ie. `-webkit-hyphens`, `-moz-align`) "private properties".

yparitcher commented:
Joining in as the target audience.
I know css, but dont use it on a day to day basis  (esp the obscure selectors) so that when i need one of them i often end up having to go to mdm. This should help with that :)

Frenzie commented:
> the person who already knows CSS? no, he already knows what to do.

The first part is indeed the target audience, but the second part isn't quite true. :-) Granted, it'll be true for the likes of `p.className` but it'd be a bit weird to leave those out.

Even for me it's neat and helpful to have an overview of the crengine prefixed/vendor-specific properties. Incidentally, those are common ways of referring to it. In-house is not though I agree it sounds nicer. See [here](https://www.w3.org/TR/CSS/#vendor-prefix) for the specification.

Commodore64user commented:
> Joining in as the target audience.

To be fair, everyone who cares enough to use GitHub is the target audience (i.e the nerds 🤓, we all are). Point is then, @poire-z  there is no need to baby proof any of it. If this is an advanced feature, man it technical. Assume people know what things are, no hand holding.

> Incidentally, those are common ways of referring to it. In-house is not though I agree it sounds nicer. See [here](https://www.w3.org/TR/CSS/#vendor-prefix) for the specification.

If that is what it is called, then I stand corrected. 

> Your wording is nice (I would remove the _Don't forget that you can_ because, even if long-press is common for that, it feels like a new information in that context :)). Fine with replacing that intro text with your wordings.

I wouldn't have added it to be begin with, see my point about 'no hand holding', I just rewrote the copy you already had.

poire-z commented:
Worth PR'ing these more pretty English changes to the main help_text text ?
```diff
--- a/frontend/apps/reader/modules/readerstyletweak.lua
+++ b/frontend/apps/reader/modules/readerstyletweak.lua
@@ -786,9 +786,7 @@ local BOOK_TWEAK_INPUT_HINT = T([[
 local CSS_SUGGESTIONS = {
     { _("Long-press for info â<93><98>"), _([[
-This popup menu provides a quick CSS cheat sheet, showing common selector syntax and classic CSS properties.
-It also shows some KOReader-specific non-standard CSS features, that can be useful with e-books.
+This menu provides a non-exhaustive CSS syntax and properties list. It also shows some KOReader-specific non-standard CSS features that can be useful with e-books.

-Most of these CSS bits are used by our categorized Style tweaks in the top menu: look there and long-press on a style tweak to see its CSS code and a description of what it does.
-If these are not enough to trigger what you want, you may need to adapt them: tap on their CSS code to copy it to the clipboard. You can then paste it here and edit it.
+Most of these bits are already used by our categorised 'Style tweaks' (found in the top menu). Long-press on any style-tweak option to see its code and its expected results. Should these not be enough to achieve your desired outcome, you may need to adjust them slightly: tap once on the CSS code-box to copy the code to the clipboard, paste it here and edit it.

 Long-press on any item in this popup to get more information on what it does and what it can help solving.
```
or better not bother translators who already translated it?

Btw, is there a way to see from a single _("xyz") in the code in which and how many languages it is current translated ?

Frenzie commented:
You can search for xyz on Weblate or in the translations repo.

Btw, our style is categorized. Categorised is for the British English translation. ^^

Commodore64user commented:
> Btw, our style is categorized. Categorised is for the British English translation. ^^

bad spelling, that is what it is. since when English is known as 'British English' and why does it need a translation? 🤓

poire-z commented:
> You can search for xyz on Weblate

OK, found ~15 translations of this one.

Frenzie commented:
Changing it isn't to be avoided or anything; we just shouldn't do it on a whim.

poire-z commented:
Let me rephrase my question then: is this change whimsical or not ? :)

Frenzie commented:
I think it's an improvement. ;-)

Commodore64user commented:
before you do, you are missing some punctuation 

>KOReader-specific non-standard CSS features
should be `KOReader-specific, non-standard CSS features` very important comma.

Commodore64user commented:
read edit

poire-z commented:
Style is not just about look ! There are stylish ways of thinking, hiding, dying :)
Most CSS indeed have some visual effect, but some have behavioural effect (ie. trigerring footnote popup, page breaks, non-linear stuff...). I thought "outcome" was more generic (and so, true) than insisting on "look".

Commodore64user commented:
> but some have behavioural effect (ie. trigerring footnote popup, page breaks, non-linear stuff...)

still looks.

Frenzie commented:
desired looks and behavior?

poire-z commented:
What's the problem with "outcome"?

poire-z commented:
Also, is "achieving a desired look" correct ? Feels like one achieves a task, not a state/fact.

Frenzie commented:
> What's the problem with "outcome"?

Nothing for me. Outcome/result/effects, all good.



> Also, is "achieving a desired look" correct ?

At the very least people do it with hair and clothing.

Commodore64user commented:
> Also, is "achieving a desired look" correct ? Feels like one achieves a task, not a state/fact.

it is, from OED: (definition 3.a) `To succeed in gaining; to acquire or attain (a desired objective, result, etc.), esp. through effort, skill, or courage; to gain, win; to reach (an end or goal). Also occasionally with infinitive as object.`

the one that concerns us: `to reach (an end or goal)`

Commodore64user commented:
> What's the problem with "outcome"?

I don't have a problem with the word, but I am making a conscious choice of words. Again what I was saying before, Looks conveys the idea of style, form, flair etc. I am not just choosing words at random, I am writing copy.

poire-z commented:
> I am writing copy.

What does that mean? (Genuinely asking.) You're a professinal writer ?

> > To succeed in gaining; to acquire or attain (a desired objective, result, etc.), esp. through effort, skill, or courage; to gain, win; to reach (an end or goal). Also occasionally with infinitive as object.
> 
> the one that concerns us: **to reach (an end or goal)**

Any of the word in that definition would work better for me than your "look" :)
`Should these not be enough to achieve your desired objective/result/end/goal`

> still looks.

That objective/result/end/goal is indeed to be appreciated visually, by _look_'ing, but that "look" is not the objective.
Again non convinced :/ (feels like it's the 3rd of many episodes :))

Commodore64user commented:
I am not a professional copywriter but I used to work in a (highly) creative industry where those skills were needed.

I think you might be missing the richness, the word look offers in that particular case. But, I am not going to keep trying to convince you...

poire-z commented:
> I used to work in a (highly) creative industry
> I think you might be missing the richness, the word look offers

I may indeed be missing that.
But that may be a reason for not using it :)

This may also explain why our appreciations diverge that much.
I get the creativeness & richness & smoothness of your wordings. And most of our users, being book readers, may be able to appreciate it. At least native English ones, but maybe not all.
But I don't think we need that for our menus and text: we don't need creativeness and poetry there. We need precision and clear statements, not rich words that could be understood in different ways, and could led some users to read them not the right way. (`Shuffle images`, `Complications`, `achieve desired look`)

And it's even more sad for you and English, because our English strings serve as the base for translators (and other readers that don't have a UI language for their native one - so we need simple and clear English too): these translators (who also may not be that confident with English) may add their own layer of richness and poetry (and I think they are more allowed to do that than us with English) - so if their poetic/rich/vague layer is added to your poetic/rich/vague layer, the translated strings could get even further from the plain and poor but real meaning.

I don't do the French translations, but translating the 3 examples above, and not knowing much about the implementation - and having looked at some French translations, I can assure you lots of translators just translate words, they don't check at all if their translations really explain the thing, unlike you that did test things for you rewording), they might come up with:
`Jouer les images` (Play the images)
`Complications` (which exists in French, meaning "Problems")
`atteindre votre apparence désirée` (reach your desired appearance)...

(I could give in to using "look" there, not because I agree, but just because nobody will probably ever read this text :))

Frenzie commented:
I think complications is a bit different than the other examples. Shuffle might have a connotation problem though, namely that shuffling is something you normally do in advance, which is different from randomly picking a card. In other words, shuffling sounds nicer but it subtly gives the wrong idea. But shuffling cards is so common that I wouldn't expect any issues.

PS Surely it should be `mélanger` or something along those lines?

Commodore64user commented:
I don't know about translations so will not comment on that. What I do know, is that as a user I like a cohesive, polished and nicely written menu as much as the next man, I am offering to take us in that direction, whether one person or millions read it is not relevant, I don't do it for those reasons. 

I just checked and complications in French is (in the horological sense)...drumroll, complications.
wtf is melanger? sounds like 'space melange' from Dune.

poire-z commented:
You don't seem to get any of my points :/ Not even the feeling that you do understand my pov but just not share it.

> as a user I like a cohesive, polished and nicely written menu as much as the next man

Let the _nicely written_ to the book authors :)
The app does not need to be judged by its nicely written wording. But by its feature and the clarity of the menu. (And I would not really mind if some people like you despise its wording - you don't read them often -, as long as you can use it for reading.)
I understand how you found some menu not cohesive and polished - but I'm qute sure their not-nicely-written text has not confused you on the feature (except "Reclaim bar height..." :))- only the layout and organisation may have.

> I just checked and complications in French is (in the horological sense)...drumroll, complications.

Good. But what's the % of French and other users that know about "horology" ? I certainly don't, and have never seen "Complications" used for meaning anything but problems or medical complications.
It's not our menu and text role to teach English and horology to users: their role is to express what the features do, and it's ok if Shakespeare died too early to contribute to our wording :)

(We indeed "mélange un jeu de carte" - shuffle a set of game cards. And we _mélange_ various spices in dishes.)

poire-z commented:
Oh, but we have https://hosted.weblate.org/projects/koreader/koreader/en_GB/.

So, there could be for the "simpler" americans (and me) the obvious wording for dummies:
```
Use a random image from folder
Settings
achieve desired goals
```

And the rich snobish British Englishmen :) (and American ones too, if they  can bear the -sion vs -zion suffixes) could get:
```
Shuffle images
Complications
achieve desired look
```
Let the poetry to translators only :)

Frenzie commented:
> wtf is melanger? sounds like 'spice melange' from Dune.

A melange is a mixture. Something like ras el hanout is a spice melange. In the world of Dune the word is used in a different manner, presumably by (former) analogy.

<hr>

> Let the poetry to translators only :)

I would object that, ignoring the obvious spelling differences, most changes or improvements should probably be applied at the source. :-) But yes, it's free to deviate much more strongly than that.

Frenzie commented:
> I pointed out in another comment like three different submenus using the word: ‘settings’ and that is a terrible design choice.

As stated:

a. The design choice is `x settings`, not `settings`.
b. Clarity is NOT bad. Settings is clear. Complications is not. It's BORING but it's not bad.

offset-torque commented:
> **poire-z** - The app does not need to be judged by its nicely written wording. But by its feature and the clarity of the menu.

Definitely agree.

> **poire-z** - I understand how you found some menu not cohesive and polished - but I'm qute sure their not-nicely-written text has not confused you on the feature.

This should be the aim, yes.

Over-optimizing the text for English language can affect translations negatively, **Commodore64user**. It might be ok for closer languages like German, French etc. since they have enough diffusion with English. But when you cross into Slavic languages, concepts change significantly. And when you reach Chinese it's a whole different world.

For example, two sentences saying the same thing:
1. We advise you to be careful with this setting since careless adjustments may introduce instabilities to your system.
2. If you enter a wrong value here, your computer might crash.

1 is more elegant yes but there is more probability of turning into soup when translating into languages I mentioned above. And of course it is harder to parse and less direct and harder to translate and doesn't convey the urgency... 

So a plain, direct language is always a good idea.

Frenzie commented:
> only it isn't. `magnifying glass > Settings`, `page with bookmark > Settings`, `cog > screen > screensaver > Settings`, `cog > status bar > Settings` I could go on, but I think I made my point. there are examples with other words too.

There's design choice and there are things that haven't been reached by it because it's always a work in progress. :-) I believe the design is in line with your vision: a dedicated word if possible and clear, x settings otherwise, and we just disagree that "settings" is especially bad. (It does probably represent the pre-2017ish design.)

poire-z commented:
> What you don’t notice is that 1) i am not using flamboyant or difficult (or poetic as you say) language

"Complications" is not flamboyant or difficult  ?! :)
May be it is not for you. But when other people tells you it is, it's that it is, for them, so probably for other people too.

> 2) the same words are utilised **everywhere** in the menus and that is BAD. I pointed out in another comment like three different submenus using the word: ‘settings’ and that is a terrible design choice.
> [...]
> only it isn't. `magnifying glass > Settings`, `page with bookmark > Settings`, `cog > screen > screensaver > Settings`, `cog > status bar > Settings` I could go on, but I think I made my point. there are examples with other words too.

I genuinely don't understand why it is BAD. Please explain why and how it is bad. (And if other people feels too it is bad, I'm ok to give in).
We don't have bread crumbs to show the above `magnifying glass > Settings`, `page with bookmark > Settings`, which I think would be fine by not repeating the object of the settings. So, OK, we could prefix it with x : `x settings>`.
But it feels it's just becoming more verbose.
We don't have breadcrumbs, but we have context, and separators:
![image](https://github.com/koreader/koreader/assets/24273478/0a4808d8-c42a-4b84-ba40-9b0775287f92)
This `Settings>` is obviously about the 2 items above, Toc and bookmarks.

I dont think this improves it at all:
![image](https://github.com/koreader/koreader/assets/24273478/48101185-c241-4ce6-93e5-a1143aaeac53)

> It seems like the menus are treated like something that gets in the way of coding cool features and not given the importance they deserve

_mildly offended_ too :)
I give them importance, in my PR and in reviews. It's probably the things that are the most often discussed, and where I really ask for feedback, in my own PRs (because, of course, my implementations are perfect and often don't need any discussion :))

Commodore64user commented:
> I dont think this improves it at all: ![image](https://private-user-images.githubusercontent.com/24273478/320511601-48101185-c241-4ce6-93e5-a1143aaeac53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1ODY1MjIsIm5iZiI6MTcxMjU4NjIyMiwicGF0aCI6Ii8yNDI3MzQ3OC8zMjA1MTE2MDEtNDgxMDExODUtYzI0MS00Y2U2LTkzZTUtYTExNDNhYWVhYzUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDE0MjM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwOGNmNTJjNDUzYzVhNmJjYzE5NzE1MThmNzFlZmY2ODliMDhjZTM2MzA2MDRmN2MyMDYwNjBlYjBlM2E5OGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U-PVOT1PKrA_HIo-0WAILKYkIVnTMHl_5_6L9jEa-z8)
 
that is silly, tbh i think `settings` should be used once. not even `x settings`. If you look at the `sleep screen` all instances of the word are gone and appropriate language has been used instead. It can be done, but it requires effort, and open mindedness to things like `complications`. Which incidentally, you only need to learn once. Next time you go to buy a Rolex you will know what they are ad will not embarrass yourself by calling them something else. 😅

> > It seems like the menus are treated like something that gets in the way of coding cool features and not given the importance they deserve
> 
> _mildly offended_ too :) I give them importance, in my PR and in reviews. It's probably the things that are the most often discussed, and where I really ask for feedback.

Again, no offence. I don't mean to single out any one person in particular, but as a whole. Also, correcting someone (particularly with long texts) takes time and effort some might not be able (or want to) spend doing. It is easier after all, to say: all good, and be done with it.

poire-z commented:
> If you look at the sleep screen all instances of the word are gone and appropriate language has been used instead.

Fine there, because you had room (or made room, as it was deep down, you could).

> that is silly, tbh i think settings should be used once. not even x settings.
> It can be done, but it requires effort, and open mindedness to things like complications. Which incidentally, you only need to learn once. 

Well, I'm totally not open to that. 1000 users will have to learn it once. Or rather, 1000 users will have to learn or 10 items once. And the User guide will need a section title "List of our fancy flamboyant menu items that all mean 'Settings'" :) And people will then really know we are using drugs and ask which one we take :) btw, what's yours?
(And we'll struggle to find alternatives to `Complications`: may be in flowerlogy, there is a term to name the thorn/spikes, which are complications to the flower petals :))

A common `x settings>` at least make it clear everywhere that this is the submenu where they will find ways to tweaks the behaviour/look of the "actions" that the other menu item in the same "separator" context launch.
In this first left menu, we don't have room - and the action to open Toc, Bookmarks, BookMap... are to be directly there - so we need a single `Toc&BM settings>` to put the many possible tweaks. We can't have `TableOfContentications>` and `Bookmarksifitudes>`

poire-z commented:
(Sorry, no mockery intended, just jokes. I feel forced to add them to keep our discussions somehow entertaining. Without them, it would feel - to me at least - unbearably dense and tense.)

Frenzie commented:
> We don't have bread crumbs to show the above `magnifying glass > Settings`, `page with bookmark > Settings`, which I think would be fine by not repeating the object of the settings. So, OK, we could prefix it with x : `x settings>`.

I meant nested instances, like Update → Settings. Whether that should be in that spot at all is an entirely different kettle of fish that offers a fairly obvious alternative win-win solution: moving those update settings somewhere in the settings menu instead (e.g., cog → Device → Updates). That's not related to the word *settings* of course, but perhaps the observed "badness" is actually something more along those lines.

poire-z commented:
^ I have no strong feelings about this in general, but I like what we have (that our olders decided, at least not me, I just followed the trend): the settings near the action.
Otherwise, we'd have a submenu somwhere
```
Settings >
  Table of contents >
  Bookmarks >
  Sleep screen>
  Updates>
```
which would indeed unclutter the main menu - but actually, the menu is not our main interface (except for non-touch users), so it does not really need simplifying.
Moreover, with out keep_menu_open tweaks, we can go tweak a setting, and just go up a menu to test the action. Having all settings moved to some darker place could need more navigations to go test it.

Frenzie commented:
Yes, it has advantages and disadvantages. I mostly like it.

Checking around a bit, some apps liberally use the word settings where appropriate (e.g., Discord), some switch around between preferences, settings and options (e.g., Slack) and some others do indeed avoid repeating the word but they can only do so because they group all settings in a settings dialog. Presumably their motivation isn't to avoid the word settings, but to group all settings together. (Discord and Slack also group their settings together ftr.) Either way they're still repeating other words, just not the word settings/options/preferences.

poire-z commented:
> > the same words are utilised everywhere in the menus and that is BAD
> I genuinely don't understand why it is BAD. Please explain why and how it is bad

Haven't got any explanation. But may be that's just about `Settings>` , and as investigated by @Frenzie , most apps have a single Settings/Preferences/Options, and some people may feel its odd to be seeing it many times ?
(Firefox uses `Customize toolbars...` for another thing not under Settings. So we could use `Customize>` if we really need some alternative to Settings or Complications...)

Anyway, I didn't notice anything wrong when I met KOReader (dunno if it had that many `Settings>` at the time).
I never really thought about it, but I had often been annoyed by these apps that have a single `Settings...` menu launching a big pane with 10 tabs with 30 checkbox each... and the need to apply & exit it for testing, and to go back to it for unchecking and chosing other options.
For usability, I find our way, with small settings submenus near the actions, a lot better.

Commodore64user commented:
Is bad because is confusing, same reason why you would not want multiple houses with the same address. You are reading, and suddenly need that one ‘thing’ which was under settings; but wait, which one? F… this and you chuck your ereader out of the window. 

Don’t judge a book by its cover, wait till the status bar menu is complete and then you can comment. But if the sleep screen is anything to go by, that is the sort of thing you are gonna get.

poire-z commented:
(Using this analogy, it feels more like you're saying: there shouldn't be two bakeries in one village.
While I think it's perfectly fine: people can refer to "the bakery near the church" or "the bakery in the south".)

poire-z commented:
> About the word "complication":
I couldn't see the beginning of the related specific discussion but we are talking about this word in jest, right?

Our discussions are alas spread among 3 issues, and interconnected indeed :). The 3rd one where `Complications` in introduced is: https://github.com/koreader/koreader/discussions/11605

offset-torque commented:
Thanks for the pointer **poire-z**. I see that it is a serious proposal. 

@Commodore64user 

I understand this concept really makes sense to you but we can not ship your brain with KOReader. I am pretty sure none of the 200+ devs here will think the same thing as you when they see this word. And I am not even talking about the multi-cultural userbase. If you are aware of this fact and still saying **"No big deal, users can learn a word or two"**, this violates one of the most important principles of UI design.

If you think that devs here are being close minded about this, I am quoting the bibles of the user interface design below:

**The Windows Interface Guidelines**
Avoid wordy descriptions for command names or messages. Irrelevant or verbose phrases clutter your design, making it difficult for users to easily extract essential information.

**Apple Developer Documentation Human Interface Guidelines**
Be clear. Choose words that are easily understood and convey the right thing.

Commodore64user commented:
of course I am being serious, what I have to say about this is, most of you (not all) seem to be very quick to judge what they haven't even seen. I think it is fair to say, that people were also a little bit unsure about the 'sleep screen' until they started playing with it and, realised how much better it was (given the constraints of course).

I didn't add an equivalent `complications` there because it wasn't necessary. I am not just making stuff up for a laugh. It really would work here (status bar) and as I said before, `items` is no better really.

as for the 'bibles', we (KOR) are already breaking a lot of the ten commandments, and so are they (apple, Microsoft). I f*ing hate the new macOS system settings menu.

offset-torque commented:
> I f*ing hate the new macOS system settings menu.

Welcome to the club of frustrated users. This is what happens when someone ignores these bibles. Unlike the real bible, these documents didn't descend from the sky, they are built on experiments and experience. That's why, when someone violates some fundamental rule, it is guaranteed that users will get emotional about the authors of the software in the spirit of **"wtf these idiots were thinking!!!"**.

I have seen so many examples of this:
- In Google's and Microsoft's web conferencing products, users can't be sure if their microphone is muted, because these trillion dollar companies ignore button and indicator guidelines to **"make the button more in line with the modern interface"**.
- Senior programmers who are in the Apple ecosystem for years saying things like **"Wow I didn't know you can do that on an iPhone"** because Apple decided to put the feature behind an obscure gesture or non-standard UI interaction.

They can afford to do this because their designers and coders don't wander the forums, or reply to GH issues to help these confused users. They are pretty isolated from the consequences of their hubris. Devs here don't have that luxury.

> I didn't add an equivalent complications there because it wasn't necessary. I am not just making stuff up for a laugh. It really would work here (status bar) and as I said before, items is no better really.

And I am saying why stop with this. Here are my proposals:

- Instead of **Library** or **Home folder** I propose **Tsundoku** which means **"the stack of books you've purchased but haven't yet read"** in Japanese, which is a perfect description of that folder and it sounds even cooler.

- Instead of **e-ink** throughout the program, we can use the word **shuǐmòhuà**, the traditional Chinese ink and wash painting technique which creates elegant images that is very similar to the display of e-ink screens. And the word is surely more elegant than the industrial "e-ink".

Do you see where this is going? Once you open the flood gates there is no stopping. How can you stop me? Will you say **"Hey these are very obscure words. It might make sense to you but no one knows them."** and I will point my finger to your cool word. Every argument you use to defend "complications" can be used by me to defend "tsundoku". 

I can even say: **"Hey since we added this, lets use the original word 積ん読. Don't worry, people will quickly get used to it and won't be confused because it is the only Japanese word in the UI".** You can't stop me because you got rid of the **"No weird words in the UI"** sign on the wall that you can point to in situations like this.

> we (KOR) are already breaking a lot of the ten commandments

And you are frustrated about it by reacting with pretty colorful words. So you are aware of what a bad UI experience is and how it makes users frustrated. But somehow you think these rules don't apply to your ideas. Probably this is the mindset of the team in Apple who designed the new macOS system settings menu.

Commodore64user commented:
> > I f*ing hate the new macOS system settings menu.
> 
> Welcome to the club of frustrated users. This is what happens when someone ignores these bibles. Unlike the real bible, these documents didn't descend from the sky, they are built on experiments and experience. That's why, when someone violates some fundamental rule, it is guaranteed that users will get emotional about the authors of the software in the spirit of **"wtf these idiots were thinking!!!"**.
> 
> I have seen so many examples of this:
> 
> * In Google's and Microsoft's web conferencing products, users can't be sure if their microphone is muted, because these trillion dollar companies ignore button and indicator guidelines to **"make the button more in line with the modern interface"**.
> * Senior programmers who are in the Apple ecosystem for years saying things like **"Wow I didn't know you can do that on an iPhone"** because Apple decided to put the feature behind an obscure gesture or non-standard UI interaction.
> 
> They can afford to do this because their designers and coders don't wander the forums, or reply to GH issues to help these confused users. They are pretty isolated from the consequences of their hubris. Devs here don't have that luxury.
> 
> > I didn't add an equivalent complications there because it wasn't necessary. I am not just making stuff up for a laugh. It really would work here (status bar) and as I said before, items is no better really.
> 
> And I am saying why stop with this. Here are my proposals:
> 
> * Instead of **Library** or **Home folder** I propose **Tsundoku** which means **"the stack of books you've purchased but haven't yet read"** in Japanese, which is a perfect description of that folder and it sounds even cooler.
> * Instead of **e-ink** throughout the program, we can use the word **shuǐmòhuà**, the traditional Chinese ink and wash painting technique which creates elegant images that is very similar to the display of e-ink screens. And the word is surely more elegant than the industrial "e-ink".
> 
> Do you see where this is going? Once you open the flood gates there is no stopping. How can you stop me? Will you say **"Hey these are very obscure words. It might make sense to you but no one knows them."** and I will point my finger to your cool word. Every argument you use to defend "complications" can be used by me to defend "tsundoku".
> 
> I can even say: **"Hey since we added this, lets use the original word 積ん読. Don't worry, people will quickly get used to it and won't be confused because it is the only Japanese word in the UI".** You can't stop me because you got rid of the **"No weird words in the UI"** sign on the wall that you can point to in situations like this.
> 
> > we (KOR) are already breaking a lot of the ten commandments
> 
> And you are frustrated about it by reacting with pretty colorful words. So you are aware of what a bad UI experience is and how it makes users frustrated. But somehow you think these rules don't apply to your ideas. Probably this is the mindset of the team in Apple who designed the new macOS system settings menu.

I agree with most of what you are saying here, just with a few caveats. First, the 'bible' also didn't descend from the sky. It was written by humans, much like the bibles of UI.

Second and more important perhaps, is the fact that I am not really changing the name of anything much like your [silly] examples suggest. I need a name for the group of 'things' which currently doesn't have one, they are currently referred as `items` but that is sort of non-commital. I am suggesting one `complications` since it has a similar use in horology. But by no means if anyone has a better suggestion; I am all ears. The point is, to reorganise this properly, I need a name for the blasted 'things', which again, currently does not exist.

poire-z commented:
> they are currently referred as `items`

`Elements >` ? Somehow used in chemistry :) but at least, it's a common concept and works as meaning "things"/"items".
`Components >` ?

Frenzie commented:
The word you're looking for is indicators. ;-)

Commodore64user commented:
> `Elements >` ? Somehow used in chemistry :) but at least, 

wouldn't users start thinking their e-readers are going to start spewing oxygen, hydrogen or helium? (sorry that was a joke, I can be silly too)

`components` sounds like a physical part, not a fan.
`elements` I don't hate it, which is good I guess.
`indicators` not sure how I feel about it.

Frenzie commented:
Elements is identical to items. Both are standard generic words. Put another way, they're just a fancy way of saying "thing." Which is not a bad thing, but all that applies to one applies to the other. Therefore items is preferable because it's what we happen to have picked over elements. Elements is a STRONG veto.

Commodore64user commented:
> Elements is identical to items. Both are standard generic words. Put another way, they're just a fancy way of saying "thing." Which is not a bad thing, but all that applies to one applies to the other. Therefore items is preferable because it's what we happen to have picked over elements. Elements is a STRONG veto.
> 
> Edit: same for components.

I prefer `complications` over `items`, and am happy to comply with your vetting regarding elements and components.

offset-torque commented:
> First, the 'bible' also didn't descend from the sky. It was written by humans, much like the bibles of UI.

No need to offend anyone while making my point about the user interfaces.

>  I am not really changing the name of anything much like your **[silly]** examples suggest.

At this point, I am thinking that you can't assess your own ideas objectively. You are too biased to accept that they can be problematic. Which makes it impossible to convince you with any kind of rational argument or objective data. From now on (in other topics too), I will not direct any arguments to you since it is an exercise in futility and waste of time for both of us.

---

Ok folks, lets see how **Moon Reader** on Android handled this status bar thingies problem. Since it has:
- Millions of users (Google Play Store stops counting after 10 million so exact number is a bit more I guess)
- It is constantly upgraded for 13 years following users' suggestions

(**SPOILER:** They dumped everything in the **Mini Status Bar** sub-menu just like KOReader. Even gestures are there. And I guess, it just works™.)

![moon_reader_status](https://github.com/koreader/koreader/assets/84420582/69471e66-13a6-4939-8620-765118696770)

Commodore64user commented:
> > I am not really changing the name of anything much like your **[silly]** examples suggest.
> 
> At this point, I am thinking that you can't assess your own ideas objectively. You are too biased to accept that they can be problematic. Which makes it impossible to convince you with any kind of rational argument or objective data. From now on (in other topics too), I will not direct any arguments to you since it is an exercise in futility and waste of time for both of us.

common man, you were literally saying, let's change the name of X to Chinese, or Y to Japanese. If calling that 'silly', is me not being objective then, I guess I'm guilty. How does that compare to me saying, I need a word for these 'things', here is one I think would work (in the same language), and if anyone has a better idea speak up, that was the whole point of #11605. 

That is very lovely of moon reader, however KOReader has way, way more options and settings. I guess the best example to show is the sleep screen, is not perfect now, but it looks a lot better than before. And that essentially was the same, all settings dumped in just one place.

Commodore64user commented:
One more thing, and this purely for entertainment purposes. The art of naming groups of things (mainly animals) using creative words isn’t new (in english at least). I remember someone joking that if you remove the plurality out of Agatha Christie’s “murder on the orient express” all you are left with is a “crow on the orient express”. Other funny collective nouns for [some] birds are: 
A wake of buzzards
A confusion of chiffchaffs
A wisdom of owls (my favourite)
A commotion of coots
A conspiracy of ravens
An asylum of cuckoos

offset-torque commented:
**To clarify:** I am not angry or something like that and this is not personal. I just find these kind of long discussions very draining. Kudos to our devs, who are more patient than I am.

>  I really like your insights, I really do. So keep them coming.

Thank you. I really appreciate this sentiment.

poire-z commented:
> I remember someone joking that if you remove the plurality out of Agatha Christie’s “murder on the orient express” all you are left with is a “crow on the orient express”. Other funny collective nouns for [some] birds are:[...]
A wisdom of owls (my favourite)

Another proof you overestimate our users - or just me :) I didn't get these jokes at all :/
~~Can you explain (for my cultural learnings for make benefit glorious developper)?~~
OK, helped myself: https://humornama.com/community/joke/what-is-the-crow-on-the-orient-express-joke/ - might help others.)

Commodore64user commented:
> > I remember someone joking that if you remove the plurality out of Agatha Christie’s “murder on the orient express” all you are left with is a “crow on the orient express”. Other funny collective nouns for [some] birds are:[...]
> > A wisdom of owls (my favourite)
> 
> Another proof you overestimate our users - or just me :) I didn't get these jokes at all :/ ~Can you explain (for my cultural learnings for make benefit glorious developper)?~ OK, helped myself: https://humornama.com/community/joke/what-is-the-crow-on-the-orient-express-joke/ - might help others.)

No, I don’t think I overestimate users, particularly people who read and are happy to pick up a dictionary to learn new words. The way I see it, you learned something today, that’s a good thing. Was it funny?

poire-z commented:
From https://github.com/koreader/koreader/discussions/11605#discussioncomment-9103430 (stupid Github discussions, moved here for more visibility and searchability):

> download the file (I just added another commit so should be up to date), and test it. It is not done yet but I really want you to see that it works.

![image](https://github.com/koreader/koreader/assets/24273478/3be9a586-f819-4b17-b4e9-574f7268f4f2)

![image](https://github.com/koreader/koreader/assets/24273478/feae75ff-e4ab-469c-8fba-cf0729edefa6)

![image](https://github.com/koreader/koreader/assets/24273478/0c464ed3-9856-46c0-b6f8-5895afd7e487)

It would work just as well (actually better, with something that isn't read naturally as a synonym of "difficulties" :)) with `Footer items` or `Footer elements` without needing a `About complications`.

Haven't checked it all, but _everything else_ (but the word "Complications") indeed "works" and looks good/better.

Commodore64user commented:
@poire-z it was meant just for you. It is NOT done yet. Work in progress.

**Nobody** offered a better alternative.

poire-z commented:
I offered Components, Elements (if it wasn't for @Frenzie low negotiations skills :)).
In any case, the alternative to your Complications (that nobody wants, for many valid and explained reasons) it to keep using Items or Footer items (I understand you really don't like it, but compromise and give up on it if you want to get everything else moving) until one gets a brighter idea.

Frenzie commented:
> I offered Components, Elements (if it wasn't for @Frenzie low negotiations skills :)).

Either everything is items, everything is elements or everything is components. There won't be a random mix of interchangeable words in order to avoid seeing the word items twice in a different clearly differentiated context. 

A menu is a different genre from prose, where you often want to use a synonym precisely to avoid repetition. Even so, in prose you should still try to avoid coming across like you just discovered thesauri exist.

Frenzie commented:
> **Nobody** offered a better alternative.

I don't think we've made it a secret that we value clarity higher than most other concerns. I'd actually love to use complications because I find it delightful, but that is simply besides the point.

It's not clear to me what reasons you have for so vehemently rejecting obvious alternatives like status icons and symbols, status indicators, indicators, indicator icons and symbols, etc. which are standard words for the concept in anything from computer and cellphone interfaces to car dashboards.

They're:

1. probably clear even if you aren't familiar with this particular use
2. already widely used for the concept
3. more precise than the word item

Commodore64user commented:
On mobile atm, if you would all let me finish it. I’ll make the PR soon and it will all be clear and visual to everyone.

as to why not use ‘symbol’ or ‘icons’, because those words are already in use in different contexts. Like when i advised  in ‘sleep screen’ to not use ‘image’ in a different context then, something selected by the user. Indicators would have worked, but some of the ‘blasted things’ don’t indicate anything (and that is quite relevant when using such a descriptive word). So I want a word that is general yet specific and that isn’t used in a different context. 

Again, let me finish the thing and you will all be able to see it in action and context.
